sdks/python/apache_beam/transforms/environments.py

```diff
@@ -76,7 +98,8 @@ def to_runner_api(self, context):
       urn=urn,
       payload=typed_param.SerializeToString()
         if isinstance(typed_param, message.Message)
-        else typed_param.encode('utf-8') if isinstance(typed_param, str)
+        else typed_param.encode('utf-8') if isinstance(typed_param,
+                                                       (str, type(u'')))
```

Comment:
@chadrik I'm not sure what would be the canonical way of doing this? since python3 doesn't have unicode / basestring anymore, but we do want python2 unicode type string to be encoded?

Reply:
It's the responsibility of `to_runner_api_parameter` to return the same type it was registered with, in this case `bytes`

```python
@Environment.register_python_environment(python_urns.EMBEDDED_PYTHON, bytes)
```

additionally, if we're following the precedent set by other registration decorators, the only valid types that can be passed are message.Message, bytes, or None. 

I saw that `PTransform.to_runner_api` has the pattern that you're using, but I don't understand it.  To me it seems like it would only work on python3 (where `str` is actually `unicode`).   Plus, I don't see anywhere that a unicode payload is used.

Another thing to note about `PTransform.register_urn` is that it implements a special pattern that works for classes or methods:

```python

  @classmethod
  def register_urn(cls, urn, parameter_type, constructor=None):
    def register(constructor):
      if isinstance(constructor, type):
        constructor.from_runner_api_parameter = register(
            constructor.from_runner_api_parameter)
        # pylint isn't smart enough to recognize when this is used
        # on a class or a method, and will emit a no-self-warning
        # in the latter case.  Rather than suppressing this at each
        # use, we fool it here through some dynamic patching that
        # pylint will also not understand.

        @contextlib.contextmanager
        def fake_static_method():
          actual_static_method = staticmethod
          globals()['staticmethod'] = lambda x: x
          yield
          globals()['staticmethod'] = actual_static_method
        with fake_static_method():
          return staticmethod(constructor)
      else:
        cls._known_urns[urn] = parameter_type, constructor
        return staticmethod(constructor)
    if constructor:
      # Used as a statement.
      register(constructor)
    else:
      # Used as a decorator.
      return register
```

That could be a good fit, allowing us to merge `register_environment` into `register_urn`

Reply:
Updated in https://github.com/violalyu/beam/pull/1/commits/7e32a4ccdc41d176fc349537d5da31de05129277